Add gpx support

jtorresfabra opened the pull request:
This PR adds support for GPX loading. It's based off this old PR https://github.com/CesiumGS/cesium/pull/2939. 

Test files for waypoints, tracks and routes have been added, there is also a SandCastle  with a dropdown menu for different gpx files here: http://localhost:8080/Apps/Sandcastle/index.html?src=GPX.html.  ( [test it online](http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/feature/rebased-gpx/Apps/Sandcastle/index.html?src=GPX.html) )

On top of the rebasedPR we added some symbology options: clampToGround, trackColor and waypoint and track images. Also path symbols have been removed in favor of polyline symbols. The main reason is that path symbols does not support clamp to ground so it was done for the sake of consistency within the symbology options.

cesium-concierge commented:
Thank you so much for the pull request @jtorresfabra! I noticed this is your first pull request and I wanted to say welcome to the Cesium community!

The [Pull Request Guidelines](https://github.com/CesiumGS/cesium/blob/master/CONTRIBUTING.md#pull-request-guidelines) is a handy reference for making sure your PR gets accepted quickly, so make sure to skim that.

* :x: Missing `CONTRIBUTORS.md` entry.
    * Please add yourself to the [contributors](https://github.com/CesiumGS/cesium/blob/feature/rebased-gpx/CONTRIBUTORS.md) file!
* :x: Missing CLA.
    * Please send in a [Contributor License Agreement](https://github.com/CesiumGS/cesium/blob/master/CONTRIBUTING.md#contributor-license-agreement-cla) (CLA) and comment back here to let us know to check this!
* :grey_question: [CHANGES.md](https://github.com/CesiumGS/cesium/blob/feature/rebased-gpx/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

OmarShehata commented on Source/DataSources/GpxDataSource.js:
```diff
@@ -0,0 +1,1029 @@
+
+  var entity = getOrCreateEntity(geometryNode, entityCollection);
+  entity.position = position;
+  // TODO different icon support
```
There's a couple of todos/commented out code here. Can you either implement or remove these? I think in this case, with different icon support, we can just remove this. GPX doesn't seem to have a way of specifying an icon for waypoints/tracks as far as I can tell. 

In general, it's OK if we don't have 100% of the spec implemented, ideally we can document what parts aren't documented if there are significant features we ignore, similar to the KMLDataSource https://cesium.com/docs/cesiumjs-ref-doc/KmlDataSource.html?classFilter=KMLDA. I don't think this is the case here.

jtorresfabra replied:
Removed

OmarShehata commented on Source/DataSources/GpxDataSource.js:
```diff
@@ -0,0 +1,1029 @@
+  if (isTimeDynamic) {
+    entity.billboard = createDefaultBillboard();
+    var url = buildModuleUrl("Assets/Textures/maki/bicycle.png");
+    entity.billboard.image = resolveHref(url);
```
I don't think `resolveHref` is actually needed, at least in how GpxDataSource is currently implemented. It is only ever called on hard-coded links to images here. If my understanding is correct, GPX doesn't actually allow for a way to link to external resources to be loaded in the application. The only links in the spec are links that can be followed (like a link to learn more about the author of the GPX file, or about this particular track etc).  

So we can simplify this code and directly use the default icons. Then the `proxy` and `sourceUri` options wouldn't be needed either. 

To replace the icons the developer can get the entities created after loading the GpxDataSource, and replace the `.image` property of the billboards, or since there's only two types of icons we can expose constructor options for them, and maybe the color of the track too.

jtorresfabra replied:
Removed the resolveHref proxy and sourceUri, and added options to pass wptImage (waypoint) and trkImage (track) as options, also trkColor aswell.

OmarShehata commented on Source/DataSources/GpxDataSource.js:
```diff
@@ -0,0 +1,1029 @@
+ * @see {@link http://www.topografix.com/gpx.asp|Topografix GPX Standard}
+ * @see {@link http://www.topografix.com/gpx/1/1/|Topografix GPX Documentation}
+ *
+ * @demo {@link}
```
I think this should link to the Sandcastle? You can see what this looks like by running `npm run generateDocumentation` and going to http://localhost:8080/Build/Documentation/GpxDataSource.html?classFilter=GPX. I would see what other classes do.

OmarShehata reviewed (commented):
This is a great start @jtorresfabra ! I've left some notes on fixes, and I think I may have found something that may have been missed in the original PR review (whether GPX needs to resolve external URLs). A few more things:

1. Should we support clamp to terrain/height reference?

On one hand, GPX data provides elevation relative to the ellipsoid, so the position in 3D should be rendered as-is to be "correct". On the other hand, a lot of GPX data is things moving on the ground, and even the sample files in the Sandcastle aren't quite correct with Cesium World Terrain (they're slightly underground). I've noticed this with my own GPS data, that some points are just going to be underground due to inaccuracies. 

I think it shouldn't be hard to do something similar to GeoJSONDataSource where you can pass an option to clamp to the ground, or pass a height reference, so it'd allow you to clamp the waypoints/path to the ground or make them relative to the ground. The default can be `HeightReference.NONE` to leave the data as-is.

The Sandcastle example should use Cesium World Terrain. 

2. We should double check the license on the sample GPX data used here and add the license to https://github.com/CesiumGS/cesium/blob/master/LICENSE.md#example-applications.

See this comment on where the data comes from https://github.com/CesiumGS/cesium/pull/2939#issuecomment-132337200. We can consider simplifying it and using fewer GPX files from one source if you can find any ones you think work better as examples.

3. Test coverage seems low, see: http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/feature/rebased-gpx/Build/Coverage/Firefox%2080.0%20(Linux%20x86_64)/DataSources/GpxDataSource.js.html. 

I also haven't tested what happens with an invalid GPX file/if errors are reported correctly etc.

OmarShehata commented:
I've added you to Cesium GS's corporate CLA. Don't forget to add yourself to [CONTRIBUTORS.md](https://github.com/CesiumGS/cesium/blob/feature/rebased-gpx/CONTRIBUTORS.md) under Cesium GS!

jtorresfabra commented:
@OmarShehata thanks for the review!

I tried to address all the comments. But there are still some concerns:

1) About clamp to ground: I implemented an option to pass clamp to ground to billboards and polylines, but unfortunately  PathGraphics don't support clamp to ground (https://github.com/CesiumGS/cesium/issues/7133). So I'm not sure if that could be misleading for users that won't see their dynamic lines clamped to terrain. I could be missing something, though.

2) I'm going to look for better data and implement more tests.

OmarShehata commented:
> About clamp to ground: I implemented an option to pass clamp to ground to billboards and polylines, but unfortunately PathGraphics don't support clamp to ground (#7133). So I'm not sure if that could be misleading for users that won't see their dynamic lines clamped to terrain. I could be missing something, though.

Ok, this is a bit tricky. Clamping _some_ things but not everything is definitely going to be confusing. I think the best options are:

1. Nothing is clamped. Not confusing, but not ideal in practice.
2. Just use polyline instead of path. So you'll see the full line and we give up the dynamic path, and just see the marker moving along that line. 

My preference is (2). It is nice to have the animated path, but it seems more useful to be able to correctly see the full path on the ground in many cases.

jtorresfabra commented:
I changed the code to use polylines instead of paths so we always can clampToGround. 

Also I implemented more tests and added my own gpx track to the examples. I checked the other files and they are really into the public domain, no license though, but gpx code was there to copy-paste. I removed some of them not used though.
The good point about those files is that they show different features of gpx (waypoints, tracks and routes).


Do you think it is ready to change from Draft PR to "Real" PR?

OmarShehata commented:
Great!

Yeah feel free to change it to "ready for review" if you feel like it's ready. I'll take another pass before bumping it to a final reviewer.

jtorresfabra commented:
Tests are failing, but master is also failing for me.
cc @OmarShehata

OmarShehata commented:
Looks great @jtorresfabra ! I pushed some minor cleanup things. I have just a few more things minor things. Once these are cleaned up please bump to Matt. It would be good to update the PR description to make it easier for the reviewer, with a summary of the changes and decisions we've discussed here, like that the GpxDataSource support passing in images, uses polyline instead of path etc. As well as any relevant instructions for testing (such as a link to the Sandcastle for this branch for convenience: http://cesium-dev.s3-website-us-east-1.amazonaws.com/cesium/feature/rebased-gpx/Apps/Sandcastle/index.html?src=GPX.html). 

Also, I noticed you force pushed a few times since the last review. It would have been good to retain the commit history so we could see the initial rebase commit followed by the commit(s) where we changed the implementation from the initial PR. Something to keep in mind for the future.

* The GpxDataSource doc page doesn't seem to document `clampToGround` ?
* The doc page for [.load](http://localhost:8080/Build/Documentation/GpxDataSource.html?classFilter=GPXD#load) says "An object with the following properties:" but doesn't list any properties?
* For the Sandcastle, should the last "GPX - Route" also have a clampToGround: true option?
* I think the `viewer.homeButton.viewModel.command();` call in each one isn't necessary and causes an extra camera flyTo
* It would be nice if one of the GPX files added in the Sandcastle shows how to use one of the options, like how to change the track color

jtorresfabra commented:
Thanks @OmarShehata! 

I think it is now ready for Matt review. 

CC: @mramato please have a look when you have the time. Thanks in advance.

eleu commented:
We are also very interested in this feature !

eleu commented:
Do you have any timeline / plan to review / merge it? Thanks a lot in advance!

cesium-concierge commented:
Thanks again for your contribution @jtorresfabra! 

No one has commented on this pull request in 30 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 30 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

cesium-concierge commented:
Thanks again for your contribution @jtorresfabra! 

No one has commented on this pull request in 30 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 30 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

cesium-concierge commented:
Thanks again for your contribution @jtorresfabra! 

No one has commented on this pull request in 30 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 30 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

cesium-concierge commented:
Thanks again for your contribution @jtorresfabra! 

No one has commented on this pull request in 30 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 30 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

cesium-concierge commented:
Thanks again for your contribution @jtorresfabra! 

No one has commented on this pull request in 90 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 90 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

cesium-concierge commented:
Thanks again for your contribution @jtorresfabra! 

No one has commented on this pull request in 90 days. Maintainers, can you review, merge or close to keep things tidy?

 _I'm going to re-bump this in 90 days. If you'd like me to stop, just comment with `@cesium-concierge stop`. If you want me to start again, just delete the comment._

sanjeetsuhag commented:
@eleu This PR fell off our radar but we're close to getting this feature in. We'd appreciate any feedback or sample data you may have.

ggetz commented on Source/DataSources/GpxDataSource.js:
```diff
@@ -0,0 +1,1022 @@
+    deferred.reject(reader.error);
+  });
+  reader.readAsText(blob);
+  return deferred;
```
`KmlDataSource` returns `deferred.promise` in a similar function. While the deferred object works in most cases I believe they don't have all of the same methods as `always`.

ggetz commented on Source/DataSources/GpxDataSource.js:
```diff
@@ -0,0 +1,1022 @@
+    links[i].setAttribute("target", "_blank");
+  }
+
+  const background = Color.WHITE;
```
To be consistant with KML, I believe the background and foreground should default to `entity.balloonStyle.bgColor` and `entity.balloonStyle.textColor`, if available.

mramato replied:
Yeah, the whole balloon thing is custom to KML. GPX has no such visual component so this is just us looking in the file and trying to generate a default description.

Honestly, all of this might be overkill and we don't actually need to generate an Entity description at all.  Depends on what use cases we are trying to support.

ggetz replied:
Got it. My misunderstanding from cross referencing KmlDataSource`.

ggetz reviewed (commented):
Overall this is looking great to me! It would be great if we could have a member of the community who's familiar with GPX take a look at this for the sake of completeness.

Not a blocker to this PR, but I'm just noting that this doesn't support the [style specification](http://www.topografix.com/GPX/gpx_style/0/2/), as I've seen some other plugins and libraries which do support.

@sanjeetsuhag Can you please update `CHANGES.md`?

ggetz commented:
Also please merge in `main` when you get the chance. A [new eslint rule has gone in](https://github.com/CesiumGS/cesium/pull/10067).

ggetz commented:
Thanks @sanjeetsuhag and all those involved! I think this is in a good place and we're ready to merge, exciting!
